rbac/management/role/relation_api_dual_write_handler.py

```diff
@@ -34,16 +38,116 @@ class DualWriteException(Exception):
+    UPDATE_SYSTEM_ROLE = "update_system_role"
+    DELETE_SYSTEM_ROLE = "delete_system_role"
+    CREATE_CUSTOM_ROLE = "create_custom_role"
+    UPDATE_CUSTOM_ROLE = "update_custom_role"
+    DELETE_CUSTOM_ROLE = "delete_custom_role"
+    ASSIGN_ROLE = "assign_role"
+    UNASSIGN_ROLE = "unassign_role"
+    CREATE_GROUP = "create_group"
+    UPDATE_GROUP = "update_group"
+    DELETE_GROUP = "delete_group"
+
+
+class ReplicationEvent:
+    """What tuples changes to replicate."""
+
+    type: ReplicationEventType
+    partition_key: str
+    add: list[common_pb2.Relationship]
+    remove: list[common_pb2.Relationship]
+
+    def __init__(
+        self,
+        type: ReplicationEventType,
+        partition_key: str,
+        add: list[common_pb2.Relationship] = [],
+        remove: list[common_pb2.Relationship] = [],
+    ):
+        """Initialize ReplicationEvent."""
+        self.partition_key = partition_key
+        self.type = type
+        self.add = add
+        self.remove = remove
+
+
+class RelationReplicator(ABC):
+    """Type responsible for replicating relations to Kessel Relations."""
+
+    @abstractmethod
+    def replicate(self, event: ReplicationEvent):
+        """Replicate the given event to Kessel Relations."""
+        pass
+
+
+class OutboxReplicater(RelationReplicator):
+    """Replicates relations via the outbox table."""
+
+    def __init__(self, role):
+        """Initialize OutboxReplicater."""
+        self.role = role
+
+    def replicate(self, event: ReplicationEvent):
+        """Replicate the given event to Kessel Relations via the Outbox."""
+        payload = self._build_replication_event(event.add, event.remove)
+        self._save_replication_event(payload, event.type, event.partition_key)
+
+    def _build_replication_event(self, relations_to_add, relations_to_remove):
+        """Build replication event."""
+        logger.info("[Dual Write] Build Replication event for role(%s): '%s'", self.role.uuid, self.role.name)
+        add_json = []
+        for relation in relations_to_add:
+            add_json.append(relationship_to_json(relation))
+
+        remove_json = []
+        for relation in relations_to_remove:
+            remove_json.append(relationship_to_json(relation))
+
+        replication_event = {"relations_to_add": add_json, "relations_to_remove": remove_json}
+        return replication_event
+
+    def _save_replication_event(self, payload, event_type, aggregateid):
+        """Save replication event."""
+        logger.info(
+            "[Dual Write] Save replication event into outbox table for role(%s): '%s'", self.role.uuid, self.role.name
+        )
+        logger.info("[Dual Write] Replication event: %s for role(%s): '%s'", payload, self.role.uuid, self.role.name)
+        # https://debezium.io/documentation/reference/stable/transformations/outbox-event-router.html#basic-outbox-table
+        outbox_record = Outbox.objects.create(
+            aggregatetype="RelationReplicationEvent",
```

Comment:
from https://debezium.io/documentation/reference/stable/transformations/outbox-event-router.html#basic-outbox-table:

 "Contains a value that the SMT appends to the name of the topic to which the connector emits an outbox message" 


@wcmitchell do we have convention for topic name ? Is preferred  to use lower case here like `relation-replication-event` 

Also  topic name can be overwritten by `route.topic.replacement` option on connector config. 

cc @mjholder @merlante

Reply:
We did some work a while ago to attempt to get a topic naming convention, but it's since gone a bit by the wayside. In general we use lowercase names and a very rough format of domain.service.purpose, like platform.upload.announce, platform.cyndi.dlq, or platform.playbook-dispatcher.run-hosts. It's worth noting that dot/'.' and underscore/'_' are considered the same for topic uniqueness by Kafka (...for some reason) so in general we use dots and avoid underscores where possible.